### PR TITLE
Add AI notebooks for live demo

### PR DESCRIPTION
**Pull request by jjbrosnan**
New AI notebooks for:

PyTorch
SciKit-Learn
TensorFlow

Merging this branch into the demo branch requires these Python packages to be available on the demo.

### CODE REVIEW

File: demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
+# Deephaven with PyTorch
+
+PyTorch is an open source machine learning library for Python.  It is a popular and widely used module that is based on the Torch framework for Lua.  PyTorch is used in some well-known applications, the most notable of which is Tesla's Autopilot advanced driver-assistance system.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md:3]:
```suggestion
PyTorch is an open source machine learning library for Python.  It is a popular and widely used module that is based on the Torch framework for Lua.  PyTorch is used in some well-known applications, most notably Tesla's Autopilot advanced driver-assistance system.
```

File: demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+\
+\
+In this notebook, we will use PyTorch to classify the Iris flower dataset.  This dataset contains 150 measurements of the petal length, petal width, sepal length, and sepal width of three different Iris flower subspecies: Iris-setosa, Iris-virginica, and Iris-versicolor.  The Iris flower dataset is commonly used in introductory AI/ML applications.  It is a classification problem: determine the class of Iris subspecies based off the four measurements given.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md:6]:
```suggestion
In this notebook, we use PyTorch to classify the Iris flower dataset.  This dataset contains 150 measurements of the petal length, petal width, sepal length, and sepal width of three different Iris flower subspecies: Iris-setosa, Iris-virginica, and Iris-versicolor.  The Iris flower dataset is commonly used in introductory AI/ML applications.  It is a classification problem: determine the class of Iris subspecies based on the four measurements given.
```

File: demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+# Additional required imports
+import random, threading, time
+import pandas as pd
+import numpy as np
+```
+
+## Import the data
+
+We will now import our data into Deephaven as a Pandas DataFrame and as a table.  The Iris dataset is available in many different places.  We'll grab it from a CSV file at a URL in our Examples repository.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md:31]:
```suggestion
Let's import our data into Deephaven as a Pandas DataFrame and as a table.  The Iris dataset is available in many different places.  We'll grab it from a CSV file at a URL in our Examples repository.
```

File: demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            print(f"Epoch: {epoch} Loss: {loss}")
+
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+```
+
+## Define how to make predictions with the trained model
+
+When our model is trained, we then need to use the trained network on the test dataset.  Let's construct a function to do just that.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md:117]:
```suggestion
After our model is trained, we can use the trained network on the test dataset.  Let's construct a function to do just that.
```

File: demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    return torch.from_numpy(np.squeeze(gather.table_to_numpy_2d(rows, columns, dtype = int)))
+
+# A function to scatter integer model predictions back into a table
+def tensor_to_table_integer(predictions, index):
+    return int(data[idx])
+```
+
+## Train the model
+
+With that done, it's time to put everything together.  Let's start by training the neural network on our training table.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md:154]:
```suggestion
With that done, it's time to put everything together.  Let's start by training the neural network on our Iris flowers table.
```

File: demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    model_func = train_model,
+    inputs = [learn.Input(["SepalLengthCM", "SepalWidthCM", "PetalLengthCM", "PetalWidthCM"], table_to_tensor_double), learn.Input("Class", table_to_tensor_integer)],
+    outputs = None,
+    batch_size = iris_train.intSize()
+)
+```
+
+## Test the trained model
+
+Testing the trained model works very similarly to training the model.  We'll substitute out the `modeel_func`, and this time, we don't need the second `learn.Input`.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md:168]:
```suggestion
Testing the trained model works very similarly to training the model.  We'll substitute out the `model_func`, and this time, we don't need the second `learn.Input`.
```

File: demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        table_writer.logRow(sepal_length, sepal_width, petal_length, petal_width)
+        time.sleep(1)
+
+thread = threading.Thread(target = write_faux_iris_measurements)
+thread.start()
+```
+
+## Use the trained model on the live data
+
+Now we've got some faux live incoming measurements.  We can just use the model we've already trained with the functions we've already created to do live classification!
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/00 Deephaven with PyTorch.md:221]:
```suggestion
Now we've got some faux live incoming measurements.  We can use the trained model on our functions to do live classification!
```

File: demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+# Additional required imports
+import random, threading, time
+import pandas as pd
+import numpy as np
+```
+
+## Import the data
+
+We will now import our data into Deephaven as a Pandas dataframe and as a table.  The Iris dataset is available at the URL in the `read_csv` line of the code blow, along with a variety of other places (even SciKit-Learn has it built in).
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md:29]:
```suggestion
Let's import our data into Deephaven as a Pandas DataFrame and as a table.  The Iris dataset is available at the URL in the `read_csv` line of the code below, along with a variety of other places (even SciKit-Learn has it built in).
```

File: demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    model_func = fit_knn,
+    inputs = [learn.Input(["SepalLengthCM", "SepalWidthCM", "PetalLengthCM", "PetalWidthCM"], table_to_numpy_double), learn.Input("Class", table_to_numpy_integer)],
+    outputs = None,
+    batch_size = iris_train.intSize()
+)
+```
+
+## Test the fitted model
+
+We've got a fitted classifier now.  Let's test it out on our test table.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md:123]:
```suggestion
We've got a fitted classifier now.  Let's try it on our test table.
```

File: demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+iris_test_knn = learn.learn(
+    table = iris_test,
+    model_func = use_fitted_knn,
+    inputs = [learn.Input(["SepalLengthCM", "SepalWidthCM", "PetalLengthCM", "PetalWidthCM"], table_to_numpy_double)],
+    outputs = [learn.Output("PredictedClass", numpy_to_table_integer, "int")],
+    batch_size = iris_test.size()
+)
+```
+
+The K-Nearest neighbors classifier worked great!  We just did some classifications using a simple model on a static data set.  That's kind of cool, but this data set is neither large or real-time.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md:135]:
```suggestion
The K-Nearest neighbors classifier worked great!  We just did some classifications using a simple model on a static data set.  That's kind of cool, but this data set is neither large nor real-time.
```

File: demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+```python
+min_petal_length, max_petal_length = raw_iris["PetalLengthCM"].min(), raw_iris["PetalLengthCM"].max()
+min_petal_width, max_petal_width = raw_iris["PetalWidthCM"].min(), raw_iris["PetalWidthCM"].max()
+min_sepal_length, max_sepal_length = raw_iris["SepalLengthCM"].min(), raw_iris["SepalLengthCM"].max()
+min_sepal_width, max_sepal_width = raw_iris["SepalWidthCM"].min(), raw_iris["SepalWidthCM"].max()
+```
+
+## Write the faux measurements to a table in real-time
+
+With these quantities now calculated and stored in memory, we need to set up alive table that we can write measurements to.  To keep it simple, we'll write measurements to the table once per second for a minute.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md:151]:
```suggestion
With these quantities now calculated and stored in memory, we need to set up a live table that we can write measurements to.  To keep it simple, we'll write measurements to the table once per second for a minute.
```

File: demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+iris_classifications_live = learn.learn(
+    table = live_iris, 
+    model_func = use_fitted_knn,
+    inputs = [learn.Input(["SepalLengthCM", "SepalWidthCM", "PetalLengthCM", "PetalWidthCM"], table_to_tensor_double)],
+    outputs = [learn.Output("PredictedClass", tensor_to_table_integer, "int")],
+    batch_size = 5
+)
+```
+
+And there we have it.  Our classifier is working on live data.  The only extra work we needed to make that happen was to set up the real-time data stream.  Pretty simple!  This may be a toy problem, but the steps hold true to more complex ones.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md:189]:
```suggestion
And there we have it.  Our classifier is working on live data.  The only extra work required was to set up the real-time data stream.  Pretty simple!  This may be a toy problem, but the steps hold true to more complex ones.
```

Comment (margaretkennedy) [reply demo/web/src/main/notebooks/Python_AI/01 Deephaven with SciKit-Learn.md:189]: should change above, too

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+# Additional required imports
+import random, threading, time
+import pandas as pd
+import numpy as np
+```
+
+## Import the data
+
+We will now import our data into Deephaven as a Pandas DataFrame and as a table.  The Iris dataset is available in many different places.  We'll grab it from a CSV file at a URL in our Examples repository.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:31]:
```suggestion
Let's import our data into Deephaven as a Pandas DataFrame and as a table.  The Iris dataset is available in many different places.  We'll grab it from a CSV file at a URL in our Examples repository.
```

Comment (margaretkennedy) [reply demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:31]: link to repository

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+```python
+def train_model(x_train, y_train):
+    model.compile(optimizer = tf.keras.optimizers.Adam(learning_rate = 0.01), loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True), metrics = ["accuracy"])
+    model.fit(x = x_train, y = y_train, epochs = 100)
+```
+
+## Define how to make predictions with the trained model
+
+When our model is trained, we will need a function to use the trained network on the testing dataset.  Let's make a function to do that.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:91]:
```suggestion
After our model is trained, we need a function to use the trained network on the testing dataset.  Let's make a function to do that.
```

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+def use_trained_model(x_test):
+    if x_test.ndim == 1:
+        x_test = np.expand_dims(x_test, 0)
+    predictions = model.predict(x_test)
+    return np.array([np.argmax(item) for item in predictions])
+```
+
+## Define the interchange between Deephaven tables and NumPy arrays
+
+There's one last step we need to take before we do the machine learning.  We need to define how our model and the two functions will interact with data in our tables.  There will be two functions that gather data from a table, and one that scatters data back into an output table.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:103]:
```suggestion
There's one last step before we do the machine learning.  We need to define how our model and the two functions will interact with data in our tables.  Two functions gather data from a table, and one scatters data back into an output table.
```

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+iris_test_predictions = learn.learn(
+    table = iris_test,
+    model_func = use_trained_model,
+    inputs = [learn.Input(["SepalLengthCM", "SepalWidthCM", "PetalLengthCM", "PetalWidthCM"], table_to_numpy_double)],
+    outputs = [learn.Output("PredictedClass", numpy_to_table_integer, "int")],
+    batch_size = iris_test.intSize()
+```
+\
+\
+The model works like a charm!  We classified the Iris flower dataset using Deephaven tables and TensorFlow.  That's kind of cool, but it's cooler to do classifications on a live feed of incoming observations.  We'll now show how to do that.  As you'll see, it turns out it's incredibly easy to do so with Deephaven!
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:147]:
```suggestion
The model works like a charm!  We classified the Iris flower dataset using Deephaven tables and TensorFlow.  That's kind of cool, but it's cooler to do classifications on a live feed of incoming observations.  We'll now show how to do that.  As you'll see, it's incredibly easy with Deephaven!
```

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+## Define bounds for making realistic faux Iris measurements
+
+We first need to set up a table that we can write faux Iris measurements in real-time to.  We can use the minimum and maximum of each measurement to make sure the faux measurements are realistic.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:151]:
```suggestion
We first need to set up a table to write faux Iris measurements in real time.  We can use the minimum and maximum of each measurement to make sure the faux measurements are realistic.
```

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+```python
+min_petal_length, max_petal_length = raw_iris["PetalLengthCM"].min(), raw_iris["PetalLengthCM"].max()
+min_petal_width, max_petal_width = raw_iris["PetalWidthCM"].min(), raw_iris["PetalWidthCM"].max()
+min_sepal_length, max_sepal_length = raw_iris["SepalLengthCM"].min(), raw_iris["SepalLengthCM"].max()
+min_sepal_width, max_sepal_width = raw_iris["SepalWidthCM"].min(), raw_iris["SepalWidthCM"].max()
+```
+
+## Write the faux measurements to a table in real-time
+
+With these quantities now calculated and stored in memory, we need to set up a live table that we can write faux measurements to.  To keep it simple, we'll write measurements to the table once per second for a minute.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:162]:
```suggestion
With these quantities now calculated and stored in memory, we need to set up a live table.  To keep it simple, we'll write measurements to the table once per second for a minute.
```

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+        table_writer.logRow(sepal_length, sepal_width, petal_length, petal_width)
+        time.sleep(1)
+
+thread = threading.Thread(target = write_faux_iris_measurements)
+thread.start()
+```
+
+## Use the trained model on live data
+
+Now we've got some faux live incoming measurements.  We can just use the model we've already trained with the functions we've already created to do live classification!
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:188]:
```suggestion
Now we've got some faux live incoming measurements.  We can now use our trained model with our functions to do live classification!
```

File: demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+iris_classifications_live = learn.learn(
+    table = live_iris,
+    model = use_trained_model,
+    inputs = [learn.Input(["SepalLengthCM", "SepalWidthCM", "PetalLengthCM", "PetalWidthCM"], table_to_numpy_double)],
+    outputs = [learn.Output("PredictedClass", numpy_to_table_integer, "int")],
+    batch_size = 5
+)
+```
+
+And that's all there is to it.  The only extra work we needed to do in order to use our model on real-time data is set up the real-time data itself.  This might be a simple problem, but this holds true for more complex problems.
```
Comment (margaretkennedy) [inline demo/web/src/main/notebooks/Python_AI/02 Deephaven with TensorFlow.md:200]:
```suggestion
That's all there is to it.  The only extra work required is to set up the real-time data itself.  This might be a simple problem, but this holds true for more complex problems.
```

**Comment by jjbrosnan** [comment]
This PR is being closed since these demo notebooks now exist in the demo branch with the correct names.
